controller/project/manager.py

```diff
@@ -219,12 +229,182 @@ def __get_crowd_label_is_data(project_id: str, is_id: str) -> Dict[str, str]:
+
+    if project_id in projects_updating:
+        return GatesIntegrationData(status=enums.GatesIntegrationStatus.UPDATING.value)
+
+    missing_tokenizer = not __tokenizer_pickle_exists(project_item.tokenizer)
+    missing_embeddings = __get_missing_embedding_pickles(project_id)
+    missing_information_sources = __get_missing_information_source_pickles(project_id)
+
+    status = enums.GatesIntegrationStatus.READY.value
+    if (
+        missing_tokenizer
+        or len(missing_embeddings) > 0
+        or len(missing_information_sources) > 0
+    ):
+        status = enums.GatesIntegrationStatus.NOT_READY.value
+
+    return GatesIntegrationData(
+        status=status,
+        missing_tokenizer=missing_tokenizer,
+        missing_embeddings=missing_embeddings,
+        missing_information_sources=missing_information_sources,
+    )
+
+
+def __tokenizer_pickle_exists(config_string: str) -> bool:
+    tokenizer_path = os.path.join(
+        "/inference/tokenizers", f"tokenizer-{config_string}.pkl"
+    )
+    return os.path.exists(tokenizer_path)
+
+
+def __get_missing_embedding_pickles(project_id: str) -> List[str]:
+    missing = []
+    embedding_items = embedding.get_finished_embeddings(project_id)
+    for embedding_item in embedding_items:
+        embedding_id = str(embedding_item.id)
+        emb_path = os.path.join(
+            "/inference", project_id, f"embedder-{embedding_id}.pkl"
+        )
+        if not os.path.exists(emb_path):
+            missing.append(embedding_id)
+    return missing
+
+
+def __get_missing_information_source_pickles(project_id: str) -> List[str]:
+    missing = []
+    is_items = information_source.get_all(project_id)
+    for is_item in is_items:
+        if is_item.type != enums.InformationSourceType.ACTIVE_LEARNING.value:
+            # only active learning information sources are pickled
+            continue
+        is_id = str(is_item.id)
+        last_payload = information_source.get_last_payload(project_id, is_id)
+        if last_payload.state == enums.PayloadState.FINISHED.value:
+            al_path = os.path.join(
+                "/inference", project_id, f"active-learner-{is_id}.pkl"
+            )
+            if not os.path.exists(al_path):
+                missing.append(is_id)
+    return missing
+
+
+def update_project_for_gates(project_id: str, user_id: str) -> None:
+
+    project_item = project.get(project_id)
+    if not project_item:
+        return
+
+    global projects_updating
+    with projects_updating_lock:
+        if project_id in projects_updating:
+            return
+        projects_updating.add(project_id)
+
+    if not __tokenizer_pickle_exists(project_item.tokenizer):
+        daemon.run(request_save_tokenizer, project_item.tokenizer)
+
+    missing_emb_pickles = __get_missing_embedding_pickles(project_id)
+    for embedding_id in missing_emb_pickles:
+        __create_embedding_pickle(project_id, embedding_id, user_id)
```

Comment:
i dont know if this works since token embeddings need the token of the tokenization process before starting, right? I don't see a check for that. 

Also I noticed for the project import embedding creation that simply starting all of them will run into issues for >3 embeddings so maybe we need to queue them as well

Reply:
As the gates integration does not trigger the tokenization itself, it would be a rare case that the tokenization is running when the gates integration endpoint is triggered. Further, for token embedding, the embedding service itself waits until the tokenization is finished. I don't think that we will run into any issues.

For the embeddings, I added a queue. Only one embedding is run at the same time.

Reply:
uploading an old file, seeing "not useable in gates" & clicking on "upgrade" will create the issue for token embeddings

I think the likeleyhood might depend on the users preferences on how they "backup" data -> one way to circumvent it would be to disable the button if an embedding needs tokens & tokenization is still running